### PR TITLE
Issues/2837

### PR DESCRIPTION
**Pull request by karhohs**
The *binary* versions or dilation, erosion, opening, and closing within skimage.morphology introduced a border of *False* values around the perimeter of an image. This is in contrast to the *non-binary* versions of dilation, erosion, opening, and closing that do not introduce this artifact. Since the non-binary versions of these morphological operations still work on boolean images the binary versions were replaced.

This is related to https://github.com/CellProfiler/CellProfiler/issues/2837

### CODE REVIEW

**Comment by 0x00b1** [comment]
Hey, @karhohs! Thanks! This looks good. However, would you mind removing the comments? We don't need them. It looks like you need to change the expected values in the filterobjects tests. See the Travis CI output:

https://travis-ci.org/CellProfiler/CellProfiler/builds/266089055?utm_source=github_status&utm_medium=notification

**Comment by karhohs** [comment]
@0x00b1 I think filterobjects.py might have been using the binary_erode bug as a feature. I've been looking into the expected results and something seems off. Is there a reason [skimage.segmentation.clear_border](http://scikit-image.org/docs/dev/api/skimage.segmentation.html?highlight=border#clear-border) isn't being implemented here, line 833 in filterobjects.py?

**Comment by 0x00b1** [comment]
@karhohs No reason. It sounds like clear_borders would fix the problem.

File: cellprofiler/modules/closing.py [hunk 1]
```diff
@@ -45,29 +45,44 @@ def visible_settings(self):
...
+
             self.function = planewise_morphology_closing
+
         elif not is_strel_2d and is_img_2d:
+
             raise NotImplementedError("A 3D structuring element cannot be applied to a 2D image.")
+
         else:
+
             if x.pixel_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: This `if ... else ...` is redundant. Set `self.function = skimage.morphology.closing` regardless of data type.

File: cellprofiler/modules/closing.py [hunk 2]
```diff
@@ -45,29 +45,44 @@ def visible_settings(self):
...
         super(Closing, self).run(workspace)
 
 
 def planewise_morphology_closing(x_data, structuring_element):
+
     y_data = numpy.zeros_like(x_data)
 
     for index, plane in enumerate(x_data):
+
         if x_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Redundant `if ... else ...`. Use `skimage.morphology.closing` regardless of data type.

File: cellprofiler/modules/dilation.py [hunk 3]
```diff
@@ -39,30 +39,46 @@ def visible_settings(self):
...
+
             self.function = planewise_morphology_dilation
+
         elif not is_strel_2d and is_img_2d:
+
             raise NotImplementedError("A 3D structuring element cannot be applied to a 2D image.")
+
         else:
+
             if x.pixel_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/dilation.py]: Use `skimage.morphology.dilation` regardless of data type.

File: cellprofiler/modules/dilation.py [hunk 4]
```diff
@@ -39,30 +39,46 @@ def visible_settings(self):
...
         super(Dilation, self).run(workspace)
 
 
 def planewise_morphology_dilation(x_data, structuring_element):
+
     y_data = numpy.zeros_like(x_data)
 
     for index, plane in enumerate(x_data):
+
         if x_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/dilation.py]: Use `skimage.morphology.dilation` regardless of data type.

File: cellprofiler/modules/erosion.py [hunk 5]
```diff
@@ -39,30 +39,46 @@ def visible_settings(self):
...
+
             self.function = planewise_morphology_erosion
+
         elif not is_strel_2d and is_img_2d:
+
             raise NotImplementedError("A 3D structuring element cannot be applied to a 2D image.")
+
         else:
+
             if x.pixel_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/erosion.py]: Use `skimage.morphology.erosion` regardless of data type.

File: cellprofiler/modules/erosion.py [hunk 6]
```diff
@@ -39,30 +39,46 @@ def visible_settings(self):
...
         super(Erosion, self).run(workspace)
 
 
 def planewise_morphology_erosion(x_data, structuring_element):
+
     y_data = numpy.zeros_like(x_data)
 
     for index, plane in enumerate(x_data):
+
         if x_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/erosion.py]: Use `skimage.morphology.erosion` regardless of data type.

File: cellprofiler/modules/opening.py [hunk 7]
```diff
@@ -40,30 +40,46 @@ def visible_settings(self):
...
+
             self.function = planewise_morphology_opening
+
         elif not is_strel_2d and is_img_2d:
+
             raise NotImplementedError("A 3D structuring element cannot be applied to a 2D image.")
+
         else:
+
             if x.pixel_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/opening.py]: Use `skimage.morphology.opening` regardless of data type.

File: cellprofiler/modules/opening.py [hunk 8]
```diff
@@ -40,30 +40,46 @@ def visible_settings(self):
...
         super(Opening, self).run(workspace)
 
 
 def planewise_morphology_opening(x_data, structuring_element):
+
     y_data = numpy.zeros_like(x_data)
 
     for index, plane in enumerate(x_data):
+
         if x_data.dtype == numpy.bool:
```
Comment (mcquin) [inline cellprofiler/modules/opening.py]: Use `skimage.morphology.erosion` regardless of data type.

**Comment by mcquin** [comment]
FWIW: `scipy.ndimage.binary_x` has the same effect as `skimage.morphology.binary_x`. Likewise, `scipy.ndimage.gray_x` and `skimage.morphology.x`. This makes me think that the behavior of `skimage.morphology.binary_x` is a bug.

If the added border to a `skimage.morphology.binary_x` operation is undesired, then we should use `skimage.morphology.x` in its place. However, I would not suggest modifying the binary erosion trick in FilterObjects. It's very useful for quickly computing pixels touching the image border.

**Comment by karhohs** [comment]
@mcquin Sounds good. I'll remove the filter objects edits. Do you agree that we should use `skimage.morphology.x` over `skimage.morphology.binary_x`? When I'm working with thresholded images, I would like the behavior of `skimage.morphology.x`.

**Comment by mcquin** [comment]
Your call! If you like `skimage.morphology.x` better than `skimage.morphology.binary_x`, and you think other users will feel the similarly, then the correct behavior for us is `skimage.morphology.x`.

**Comment by karhohs** [comment]
If no one else disagrees, let's go with my preference. I'm betting it is the generally preferred behavior.

**Comment by mcquin** [comment]
Go for it, @karhohs :)

**Comment by mcquin** [comment]
Nice work, dude!

**Pull request merged**